src/compose/resolve-block-map.ts

```diff
@@ -102,7 +102,24 @@ export function resolveBlockMap(
         : composeEmptyNode(ctx, offset, sep, null, valueProps, onError)
       offset = valueNode.range[2]
       const pair = new Pair(keyNode, valueNode)
-      if (ctx.options.keepSourceTokens) pair.srcToken = collItem
+      if (ctx.options.keepSourceTokens) {
+        pair.srcToken = collItem
+
+        // Check to see if the key tokens exist and have an indentation set.
+        // If so, we can compute the difference between them and preserve it
+        // if the preserveCollectionIndentation option is set.
+        const keyIndent: number | undefined = (collItem?.key as any)?.indent
+        const valueIndent: number | undefined = (collItem?.value as any)?.indent
+
+        if (
+          ctx.options.preserveCollectionIndentation &&
+          keyIndent !== undefined &&
+          valueIndent !== undefined &&
+          valueIndent >= keyIndent
```

Comment:
TypeScript doesn't permit you to compare `number|undefined` values with `>=`, even though JavaScript has the right semantics. So we either need a cast (which seems unnecessary) or we need to explicitly include the logic to exclude `undefined` values.

Reply:
I've fixed this to avoid the `as any` casts; I didn't notice the destructuring in the first past. Now it's simplified (somewhat) to

```ts
const keyIndent: number | undefined =
  key && 'indent' in key ? key.indent : undefined
const valueIndent: number | undefined =
  value && 'indent' in value ? value.indent : undefined
```

The extra logic here is needed to convince TypeScript that it's okay:

- `key.indent` is only valid if we know that `"indent"` is in `key` in the first place (and some of the variants don't include an `indent` field), so we have to do `"indent" in key` before we can write `key.indent`
- `'indent' in key` is only valid if `key` is not null/undefined, so we have to check `key &&` first